File: reboot/examples/bank-pydantic/api/bank/v1/pydantic/bank.py
from reboot.api import API, Field, Methods, Model, Reader, Transaction, Type


class BankState(Model):
    customer_ids_map_id: str = Field(tag=1)


class SignUpRequest(Model):
    customer_id: str = Field(tag=1)


class AllCustomerIdsResponse(Model):
    customer_ids: list[str] = Field(tag=1)


class TransferRequest(Model):
    from_account_id: str = Field(tag=1)
    to_account_id: str = Field(tag=2)
    amount: float = Field(tag=3)


class OpenCustomerAccountRequest(Model):
    initial_deposit: float = Field(tag=1)
    customer_id: str = Field(tag=2)


class CustomerAccount(Model):
    account_id: str = Field(tag=1)
    balance: float = Field(tag=2)


class CustomerAccounts(Model):
    customer_id: str = Field(tag=1)
    accounts: list[CustomerAccount] = Field(tag=2)


class AccountBalancesResponse(Model):
    balances: list[CustomerAccounts] = Field(tag=1)


BankMethods = Methods(
    create=Transaction(
        request=None,
        response=None,
        factory=True,
    ),
    sign_up=Transaction(
        request=SignUpRequest,
        response=None,
    ),
    all_customer_ids=Reader(
        request=None,
        response=AllCustomerIdsResponse,
    ),
    transfer=Transaction(
        request=TransferRequest,
        response=None,
    ),
    open_customer_account=Transaction(
        request=OpenCustomerAccountRequest,
        response=None,
    ),
    account_balances=Reader(
        request=None,
        response=AccountBalancesResponse,
    ),
)

api = API(
    Bank=Type(
        state=BankState,
        methods=BankMethods,
    ),
)

File: reboot/examples/bank-zod/api/bank/v1/bank.ts
import { z } from "zod/v4";

import { reader, transaction } from "@reboot-dev/reboot-api";

export const Bank = {
  state: {
    customerIdsMapId: z.string().meta({ tag: 1 }),
  },
  methods: {
    create: transaction({
      factory: {},
      request: {},
      response: z.void(),
    }),
    signUp: transaction({
      request: {
        customerId: z.string().meta({ tag: 1 }),
      },
      response: z.void(),
    }),
    allCustomerIds: reader({
      request: {},
      response: {
        customerIds: z.array(z.string()).meta({ tag: 1 }),
      },
    }),
    transfer: transaction({
      request: {
        fromAccountId: z.string().meta({ tag: 1 }),
        toAccountId: z.string().meta({ tag: 2 }),
        amount: z.number().meta({ tag: 3 }),
      },
      response: z.void(),
    }),
    openCustomerAccount: transaction({
      request: {
        initialDeposit: z.number().meta({ tag: 1 }),
        customerId: z.string().meta({ tag: 2 }),
      },
      response: z.void(),
    }),
    accountBalances: reader({
      request: {},
      response: {
        balances: z
          .array(
            z.object({
              customerId: z.string().meta({ tag: 1 }),
              accounts: z
                .array(
                  z.object({
                    accountId: z.string().meta({ tag: 1 }),
                    balance: z.number().meta({ tag: 2 }),
                  })
                )
                .meta({ tag: 2 }),
            })
          )
          .default(() => [])
          .meta({ tag: 1 }),
      },
    }),
  },
};

export const api = {
  Bank,
};
